arc: refactor the soc folder for em_starterkit

vonhust opened the pull request:
# PART1: refactor the soc folder

The original em7d, em9d and em11d are different configurations of
em_starterkit. They have the same peripherals, e.g. uart, spi, gpio
, ddr. The differences of them are in arc core configurations, interrupt
number assignment.

So em7d, em9d and em11d can be viewed in the same SoC family or SoC
series.

Referring other arch's implementation, this commit merges em7d, em9d
and em11d into the same SoC, named snps_emsk. This will eliminate
unnecessary duplication and make it easier for future maintainment.

# PART2: optimize the dts definitions
*  add yaml files for designware gpio and designware spi
*  optimize, improve and bug fix in dts.fixup
*  optimize the dts files of em_starterkit
*  lots of bug fixes

Commit:
arch: arc: refactor the soc part of em_starterkit

The original em7d, em9d and em11d are different configurations of
em_starterkit. They have the same peripherals, e.g. uart, spi, gpio
, ddr. The differences of them are in arc core configurations, interrupt
number assignment.

So em7d, em9d and em11d can be viewed in the same SoC family or SoC
series.

Referring other arch's implementation, this commit merges em7d, em9d
and em11d into the same SoC, named snps_emsk. This will eliminate
unnecessary duplication and make it easier for future maintainment.

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

Commit:
dts: optimize the dts for em_starterkit

eliminate unnecessary duplication

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

Commit:
doc: update the doc of em_starterkit

update some content of doc of em_starterkit

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

Commit:
dts: fixes in the dts of em_starterkit em9d configuration

* em9d is configured to use ccm, and no sram
* choose dccm as virtual sram to generate the correct definitions
of CONFIG_SRAM_XXX which are used in many places, e.g. lib-hooks.c.

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

dbkinder reviewed (approved):
doc changes look OK

ruuddw reviewed (changes requested):
The refactoring simplifies and removes duplication, looks good. Few comments/suggestions:
- clean up GCC_M_CPU selection
- change 'sram0' section in the dts to dram as it's DDR memory, and add some aliases to sram if needed.
- move EMSK Core Selection from ARC options to Board options
- consider adding explicit em_starterkit_em9d_defconfig.* and including that in em_starterkit_defconfig, so all 3 variants can be easily found

ruuddw commented:
Note: this PR addessses issue #6621.

ruuddw commented on arch/arc/soc/snps_emsk/linker.ld:
```diff
@@ -1,33 +1,40 @@
  * DRAM includes the exception vector table at reset, which is at
  * the beginning of the region.
  */
+#if defined(CONFIG_SRAM_BASE_ADDRESS) && (CONFIG_SRAM_SIZE > 0)
```
What's the purpose of these #ifdefs? there are no #else with alternatives or warning/error messages.
Also, the "TODO" comment below on SRAM doesn't seem to be a TODO, or any plans to change/update?

vonhust replied:
it's used to avoid wrong definition in include\arc\arc\v2\linker.ld
If CONFIG_SRAM_BASE_ADDRESS is not defined.  without #if defined here,  no matter how, SRAM_START will be defined to CONFIG_SRAM_BASE_ADDRESS,  then region SRAM in MEMORY of linker.ld will be generated. This is a wrong definition.

Commit:
dts: optimize and bug fixes the dts of em_starterkit

* add gpio, i2c, spi definitions
* optimize and bug fix the dts.fixup
* optimize and bug fix the em_starterkit related definitions
  in dts folder

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

vonhust commented on dts/arc/emsk.dtsi:
```diff
@@ -8,6 +8,12 @@
 		compatible = "simple-bus";
 		ranges;
 
+		sram0: memory@10000000 {
```
already rename to ddr0

vonhust commented on arch/arc/soc/snps_emsk/CMakeLists.txt:
```diff
@@ -0,0 +1,20 @@
+zephyr_cc_option_ifdef(CONFIG_SOC_EMSK_EM11D -mcpu=em4_fpuda)
+zephyr_cc_option(-mno-sdata -mdiv-rem -mswap -mnorm)
+zephyr_cc_option(-mmpy-option=6 -mbarrel-shifter)
+zephyr_cc_option_ifdef(CONFIG_SOC_EMSK_EM7D --param l1-cache-size=16384)
```
use zephyr_cc_option(-mcpu=${GCC_M_CPU}) to replace.

vonhust commented on arch/arc/soc/snps_emsk/Kconfig.defconfig.em7d:
```diff
@@ -1,14 +1,14 @@
 
-config SOC
-	default em7d
+config CPU_EM4_DMIPS
```
CPU_EM4_DMIPS cannot be removed here. It was used to define what kind of CPU it is. gcc-mpu-cmake will use CPU_EM4_DMIP and so on to set GCC_M_CPU.  The definitions in gcc-mpu-cmake are used to verify toolchain.   -mcpu option is used first there.

Commit:
arch: arc: reuse GCC_M_CPU in cmake file and remove unused comments

* reuse GCC_M_CPU in cmake file
* remove unused comments in linker.ld

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

Commit:
boards: em_starterkit: add pmod mux init

* add pmod mux init to config the pmode interfaces
  to the specific functions

* without pmod mux init, all pmod interfaces will be
GPIO, no interfaces for spi/i2c0/i2c1/uart0

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

Commit:
arch: arc: add the missing SPI_DW_SPI_CLOCK in dts.fixup

* the SPI share the same clk source with uart

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

Commit:
arch: arc: explictly list all def config

it seems regular expresssion is not supported in "source xxx" after
latest changes in build system.

So explictly list all def config

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>
